### PR TITLE
Adding a query traversal for defer

### PR DESCRIPTION
**Pull request by kmoore-intuit**
# What Changed

# Why

Todo:

- [ ] Add tests
- [ ] Add docs

### CODE REVIEW

File: src/main/java/com/intuit/graphql/orchestrator/utils/MultiEIGenerator.java [hunk 1]
```diff
@@ -39,7 +43,12 @@ public Flux<ExecutionInput> generateEIs() {
                 this.timeProcessedSplit = System.currentTimeMillis();
                 //Adds elements to list of eis that need to be processed
                 try {
-                    this.eis.addAll(MultipartUtil.splitMultipartExecutionInput(emittedEI));
+                    Document rootDocument = new Parser().parseDocument(emittedEI.getQuery());
```
Comment (CNAChino) [inline src/main/java/com/intuit/graphql/orchestrator/utils/MultiEIGenerator.java:46]: I think Parser object is thread-safe.  We can have a static one.

File: src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryVisitor.java [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+import java.util.stream.Collectors;
+
+import static com.intuit.graphql.orchestrator.utils.DirectivesUtil.DEFER_DIRECTIVE_NAME;
+import static com.intuit.graphql.orchestrator.utils.DirectivesUtil.DEFER_IF_ARG;
+import static com.intuit.graphql.orchestrator.utils.IntrospectionUtil.__typenameField;
+import static graphql.util.TreeTransformerUtil.changeNode;
+
+/*
+* Purpose of class: traverse query and create forked ei whenever there is a valid defer directive
+* */
```
Comment (CNAChino) [inline src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryVisitor.java:45]: extra *

File: src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryVisitor.java [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+    //constants
+    public static final String GENERATED_EIS = "GENERATED_EIS";
+    private static final AstTransformer AST_TRANSFORMER = new AstTransformer();
+
+    //variables that will be manipulated throughout
+    private final List<ExecutionInput> generatedEIs =  new ArrayList<>();
+    private final Set<String> neededFragmentSpreads = new HashSet<>();
+
+    //variables that are set when visitor is built
+    private ExecutionInput originalEI;
```
Comment (CNAChino) [inline src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryVisitor.java:58]: I think non final vars should be moved lower to final ones

File: src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryVisitor.java [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+
+        return originalEI.transform(builder -> builder.query(query));
+    }
+
+    /*
+    * Transforms node by removing defer directive from the node.
+    * Returns transformed node
+    * Throws an exception if it is at a location that is not supported.
+    * */
+    private Node pruneDeferDirective(Node deferredNode) {
```
Comment (CNAChino) [inline src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryVisitor.java:201]: might want to generalize that is not only for Defer and move to a util

File: src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryVisitor.java [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    * Functions:
+    *    Updates the field if it contains defer directive.
+    *    If it is a valid deferred node, generate new EI
+    */
+    @Override
+    public TraversalControl visitField(Field node, TraverserContext<Node> context) {
+        if(containsValidDeferDirective(node)) {
+            Node prunedNode = pruneDeferDirective(node);
+
+            if(isValidDeferredSelection(node)) {
```
Comment (CNAChino) [inline src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryVisitor.java:77]: just curious if this validation should be run first before `pruneDeferDirective(node)`?

File: src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryVisitor.java [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+
+            return changeNode(context, prunedNode);
+        }
+
+        return TraversalControl.CONTINUE;
+    }
+
+    @Override
+    public TraversalControl visitInlineFragment(InlineFragment node, TraverserContext<Node> context) {
+        if(containsValidDeferDirective(node)) {
```
Comment (CNAChino) [inline src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryVisitor.java:110]: is there a way to move this code to a method and reuse?

File: src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryVisitor.java [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+        }
+
+        return TraversalControl.CONTINUE;
+    }
+
+    /*
+    * Checks if it is necessary to create ei for deferred field.
+    * Currently, selection should be skipped if all the children field are deferred resulting in an empty selection set.
+    */
+    private boolean isValidDeferredSelection(Node selection) {
```
Comment (CNAChino) [inline src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryVisitor.java:127]: I'm thinking of renaming the method `hasNonDeferredSelection()`

File: src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryVisitor.java [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+                //it is leaf node with no children
+        return !possibleSelectionSet.isPresent() ||
+                // at least one of the children are not deferred
+                possibleSelectionSet.get().getChildren().stream().anyMatch(child -> !containsValidDeferDirective(child));
+    }
+
+    /*
+    * Verifies that Node has defer directive that is not disabled
+    */
+    private boolean containsValidDeferDirective(Node node) {
```
Comment (CNAChino) [inline src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryVisitor.java:143]: Should we rename to containsEnabledDeferDirective?  I think a disabled defer is also valid directive.

File: src/main/java/com/intuit/graphql/orchestrator/utils/MultiEIGenerator.java (hunk 1 above)
Comment (ashpak-shaikh) [reply src/main/java/com/intuit/graphql/orchestrator/utils/MultiEIGenerator.java:46]: Also, lets see if we already have a static parser somewhere in the code. Same for AstTransformer

File: src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryVisitor.java (hunk 5 above)
Comment (kmoore-intuit) [reply src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryVisitor.java:77]: regardless of if the validation, the directive needs to be removed. The reason being, if there is a nested field later on the defer directive should be removed or else the query to the downstream will have the directive.

File: src/main/java/com/intuit/graphql/orchestrator/utils/MultiEIGenerator.java (hunk 1 above)
Comment (kmoore-intuit) [reply src/main/java/com/intuit/graphql/orchestrator/utils/MultiEIGenerator.java:46]: Located in GraphQLUtil file.. Next commit will have changes

File: src/main/java/com/intuit/graphql/orchestrator/deferDirective/DeferUtil.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+     */
+    public static boolean hasNonDeferredSelection(Node selection) {
+        Optional<SelectionSet> possibleSelectionSet = ((List<Node>)selection.getChildren())
+                .stream()
+                .filter(SelectionSet.class::isInstance)
+                .map(SelectionSet.class::cast)
+                .findAny();
+
+        //it is leaf node with no children
+        return !possibleSelectionSet.isPresent() ||
```
Comment (ashpak-shaikh) [inline src/main/java/com/intuit/graphql/orchestrator/deferDirective/DeferUtil.java:30]: why split the stream operation in two steps?

File: src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryModifier.java [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+
+    /*
+    * Generates new node
+    * Transforms the parentNode with a new selection set consisting of the pruned child and typename fields
+    * */
+    private Node constructNewPrunedNode(Node parentNode, Node prunedChild) {
+        List<Selection> selections = new ArrayList<>();
+        selections.add((Selection) prunedChild);
+
+        if(parentNode instanceof Field || parentNode instanceof FragmentDefinition || parentNode instanceof InlineFragment) {
```
Comment (ashpak-shaikh) [inline src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryModifier.java:156]: There should be a common parent for these 3

File: src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/EIAggregateVisitorModifier.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+                break;
+            }
+        }
+
+        //calls the parents original method if it was set to continue
+        return (this.visitorResult == TraversalControl.CONTINUE) ? this.visitSelection(node, context) : this.visitorResult;
+    }
+
+    public TraversalControl visitFragmentSpread(FragmentSpread node, TraverserContext<Node> context) {
+        for (NodeVisitorStub visitor: visitorStubs) {
```
Comment (ashpak-shaikh) [inline src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/EIAggregateVisitorModifier.java:58]: we should be able to move this loop into a method somehow

Comment (ashpak-shaikh) [reply src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/EIAggregateVisitorModifier.java:58]: will make this class even sweeter

File: src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/QueryVisitorStub.java [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+import graphql.util.TraverserContext;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/*
+* Skeleton Visitor class that just extends NodeVisitorStub.
+* It allows Visitors to get results and share the context between QueryVisitorStub
+* */
+public abstract class QueryVisitorStub extends NodeVisitorStub {
```
Comment (ashpak-shaikh) [inline src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/QueryVisitorStub.java:56]: this should exist in the graphql-java lib

**Comment by CNAChino** [comment]
should this PR be closed?  looks like same with https://github.com/graph-quilt/graphql-orchestrator-java/pull/154

**Pull request closed without merging**